# DS5460 Milestone 2 - EDA 

### Ingesting Files to PySpark

Author: Anne Tumlin

Date: 03/21/25

Now that we have taken the files from the original GCS bucket, extracted them, and put them in our local GCS bucket (see `docs/EXTRACTION_PROCESS` in GitHub for more details), we can begin to ingest our data into PySpark. 

In [2]:
from pyspark.sql import SparkSession

spark = SparkSession.builder \
    .appName("app_name") \
    .getOrCreate()

Don't forget to edit the path with YOUR google storage bucket here. 

In [3]:
json_path = "gs://ds5460-tumlinam-fp-bucket/gridopt-dataset-tmp/dataset_release_1/pglib_opf_case500_goc/group_1/*.json"

### Testing on a Small Scale
Let's try this with only 100 files first. 

In [9]:
from google.cloud import storage
client = storage.Client()

Reminder: Change bucket_name 

In [10]:
bucket_name = "ds5460-tumlinam-fp-bucket"
bucket = client.get_bucket(bucket_name)

In [12]:
prefix = "gridopt-dataset-tmp/dataset_release_1/pglib_opf_case500_goc/group_1/"

In [13]:
# List blobs (files) in the specified prefix and collect the first 100 file paths
blobs = bucket.list_blobs(prefix=prefix)
file_paths = []
for blob in blobs:
    file_paths.append(f"gs://{bucket_name}/{blob.name}")
    if len(file_paths) >= 100:
        break

In [14]:
print("Loading the following 100 JSON file paths:")
for path in file_paths:
    print(path)

Loading the following 100 JSON file paths:
gs://ds5460-tumlinam-fp-bucket/gridopt-dataset-tmp/dataset_release_1/pglib_opf_case500_goc/group_1/example_15000.json
gs://ds5460-tumlinam-fp-bucket/gridopt-dataset-tmp/dataset_release_1/pglib_opf_case500_goc/group_1/example_15001.json
gs://ds5460-tumlinam-fp-bucket/gridopt-dataset-tmp/dataset_release_1/pglib_opf_case500_goc/group_1/example_15002.json
gs://ds5460-tumlinam-fp-bucket/gridopt-dataset-tmp/dataset_release_1/pglib_opf_case500_goc/group_1/example_15003.json
gs://ds5460-tumlinam-fp-bucket/gridopt-dataset-tmp/dataset_release_1/pglib_opf_case500_goc/group_1/example_15004.json
gs://ds5460-tumlinam-fp-bucket/gridopt-dataset-tmp/dataset_release_1/pglib_opf_case500_goc/group_1/example_15005.json
gs://ds5460-tumlinam-fp-bucket/gridopt-dataset-tmp/dataset_release_1/pglib_opf_case500_goc/group_1/example_15006.json
gs://ds5460-tumlinam-fp-bucket/gridopt-dataset-tmp/dataset_release_1/pglib_opf_case500_goc/group_1/example_15007.json
gs://ds5460-t

**IMPORTANT NOTE:** After testing and running into issues with the schema, I discovered that due to the way the JSON files are formatted we must utilize the multiline read option. Otherwise, our data will not be read in properly. Instead, it will lead to the error `|-- _corrupt_record: string (nullable = true)`. 

In [19]:
# Make sure to use multiline read option!
df_small = spark.read.option("multiline", "true").json(file_paths)

In [20]:
df_small.printSchema()

root
 |-- grid: struct (nullable = true)
 |    |-- context: array (nullable = true)
 |    |    |-- element: array (containsNull = true)
 |    |    |    |-- element: array (containsNull = true)
 |    |    |    |    |-- element: double (containsNull = true)
 |    |-- edges: struct (nullable = true)
 |    |    |-- ac_line: struct (nullable = true)
 |    |    |    |-- features: array (nullable = true)
 |    |    |    |    |-- element: array (containsNull = true)
 |    |    |    |    |    |-- element: double (containsNull = true)
 |    |    |    |-- receivers: array (nullable = true)
 |    |    |    |    |-- element: long (containsNull = true)
 |    |    |    |-- senders: array (nullable = true)
 |    |    |    |    |-- element: long (containsNull = true)
 |    |    |-- generator_link: struct (nullable = true)
 |    |    |    |-- receivers: array (nullable = true)
 |    |    |    |    |-- element: long (containsNull = true)
 |    |    |    |-- senders: array (nullable = true)
 |    |    |  

In [21]:
df_small.show(2)

+--------------------+--------------------+--------------------+
|                grid|            metadata|            solution|
+--------------------+--------------------+--------------------+
|[[[[100.0]]], [[[...| [443934.8106702195]|[[[[[1.2271252469...|
|[[[[100.0]]], [[[...|[465533.45792886155]|[[[[[1.3700529900...|
+--------------------+--------------------+--------------------+
only showing top 2 rows



### Testing on the Full Dataset

In [ ]:
df = spark.read.option("multiline", "true").json(json_path)

In [ ]:
df.printSchema()

### Testing the Summarizing Relevant Datasets Code for the New Dataset
  
Author: Ewan Long
  
Date: 03/23/25
  
We are trying to test and revise the original code we have in the Milestone 1 as we changed our dataset (see `docs/EXTRACTION_PROCESS` in GitHub for reasons & more details), we have ingested our data into PySpark. 
  
The following code will try to aggregate data and join the key tables (see Milestone 1 Notebook and doc for reasons)

___

Firstly, initialize the SparkSession. We will process the first 100 JSON files here for faster processing and test verification.

In [1]:
from google.cloud import storage
client = storage.Client()

In [2]:
import os
import subprocess
from pyspark.sql import SparkSession
import pyspark.sql.functions as F
from pyspark.sql.functions import col, explode, input_file_name, expr, sum as spark_sum, avg, count

In [4]:
spark = SparkSession.builder \
    .appName("app_name") \
    .getOrCreate()

In [6]:
json_path = "gs://dataproc-temp-us-central1-299400297029-n5lhuci2/gridopt-dataset-tmp/dataset_release_1/pglib_opf_case500_goc/group_1/*.json"

In [7]:
# Change the bucket to your own's
bucket_name = "bucketseattledonna"
bucket = client.get_bucket(bucket_name)

In [8]:
prefix = "gridopt-dataset-tmp/dataset_release_1/pglib_opf_case500_goc/group_1/"

In [9]:
# List blobs (files) in the specified prefix and collect the first 100 file paths
blobs = bucket.list_blobs(prefix=prefix)
file_paths = []
for blob in blobs:
    file_paths.append(f"gs://{bucket_name}/{blob.name}")
    if len(file_paths) >= 100:
        break

In [10]:
print("Loading the following 100 JSON file paths:")
for path in file_paths:
    print(path)

Loading the following 100 JSON file paths:
gs://bucketseattledonna/gridopt-dataset-tmp/dataset_release_1/pglib_opf_case500_goc/group_1/example_15000.json
gs://bucketseattledonna/gridopt-dataset-tmp/dataset_release_1/pglib_opf_case500_goc/group_1/example_15001.json
gs://bucketseattledonna/gridopt-dataset-tmp/dataset_release_1/pglib_opf_case500_goc/group_1/example_15002.json
gs://bucketseattledonna/gridopt-dataset-tmp/dataset_release_1/pglib_opf_case500_goc/group_1/example_15003.json
gs://bucketseattledonna/gridopt-dataset-tmp/dataset_release_1/pglib_opf_case500_goc/group_1/example_15004.json
gs://bucketseattledonna/gridopt-dataset-tmp/dataset_release_1/pglib_opf_case500_goc/group_1/example_15005.json
gs://bucketseattledonna/gridopt-dataset-tmp/dataset_release_1/pglib_opf_case500_goc/group_1/example_15006.json
gs://bucketseattledonna/gridopt-dataset-tmp/dataset_release_1/pglib_opf_case500_goc/group_1/example_15007.json
gs://bucketseattledonna/gridopt-dataset-tmp/dataset_release_1/pglib_o

Then, read all JSON files with the multiline option enabled so that nested JSON is parsed correctly. Adding the original **filename** column for identification and join purposes.

In [11]:

df = spark.read.option("multiline", "true").json(file_paths).withColumn("filename", input_file_name())

# Print the schema to verify that 'grid.nodes.generator' and others are parsed correctly.
df.printSchema()

root
 |-- grid: struct (nullable = true)
 |    |-- context: array (nullable = true)
 |    |    |-- element: array (containsNull = true)
 |    |    |    |-- element: array (containsNull = true)
 |    |    |    |    |-- element: double (containsNull = true)
 |    |-- edges: struct (nullable = true)
 |    |    |-- ac_line: struct (nullable = true)
 |    |    |    |-- features: array (nullable = true)
 |    |    |    |    |-- element: array (containsNull = true)
 |    |    |    |    |    |-- element: double (containsNull = true)
 |    |    |    |-- receivers: array (nullable = true)
 |    |    |    |    |-- element: long (containsNull = true)
 |    |    |    |-- senders: array (nullable = true)
 |    |    |    |    |-- element: long (containsNull = true)
 |    |    |-- generator_link: struct (nullable = true)
 |    |    |    |-- receivers: array (nullable = true)
 |    |    |    |    |-- element: long (containsNull = true)
 |    |    |    |-- senders: array (nullable = true)
 |    |    |  

Then we can do the data processing. Let's process generator data first. 
  
I will explode nested JSON and extract generator attributes.

In [12]:
gen_df = df.select(
    "filename",
    explode("grid.nodes.generator").alias("generator_array")
).select(
    "filename",
    col("generator_array")[0].alias("mbase"),
    col("generator_array")[1].alias("pg"),
    col("generator_array")[2].alias("pmin"),
    col("generator_array")[3].alias("pmax"),
    col("generator_array")[4].alias("qg"),
    col("generator_array")[5].alias("qmin"),
    col("generator_array")[6].alias("qmax"),
    col("generator_array")[7].alias("vg"),
    col("generator_array")[8].alias("cost_squared"),
    col("generator_array")[9].alias("cost_linear"),
    col("generator_array")[10].alias("cost_offset")
)

gen_df.show(5) # verify the extracted attributes

+--------------------+-----+-------------------+--------------------+-----+--------+--------+-------+---+------------+-----------+-----------+
|            filename|mbase|                 pg|                pmin| pmax|      qg|    qmin|   qmax| vg|cost_squared|cost_linear|cost_offset|
+--------------------+-----+-------------------+--------------------+-----+--------+--------+-------+---+------------+-----------+-----------+
|gs://bucketseattl...| 46.9|            0.34463|             0.22026|0.469| 0.07856|-0.02298| 0.1801|1.0|         0.0|     3000.0|        0.0|
|gs://bucketseattl...|888.2|            6.43217|  3.9823399999999998|8.882| 1.27013|-0.72832|3.26858|1.0|       10.98|      910.9|        0.0|
|gs://bucketseattl...| 25.0|            0.16691| 0.08381999999999999| 0.25|0.041875|-0.01225|  0.096|1.0|         0.0|     3000.0|        0.0|
|gs://bucketseattl...| 25.0|            0.13858|0.027160000000000004| 0.25|0.041875|-0.01225|  0.096|1.0|         0.0|     3000.0|        0.0|

I will try to aggregates generator features at the snapshort level, summarizing key characteristics for each filename.
  
The following statistics will be extracted and computed:
- Total number of generators
- Sum of generated power (pg)
- Average generator voltage (vg)
- Average generator cost parameters (quadratic, linear, and offset terms)

In [13]:
# Aggregate generator features by snapshot (filename)
gen_agg = gen_df.groupBy("filename").agg(
    count("*").alias("num_generators"),
    spark_sum("pg").alias("total_pg"),
    avg("vg").alias("avg_vg"),
    avg("cost_squared").alias("avg_cost_squared"),
    avg("cost_linear").alias("avg_cost_linear"),
    avg("cost_offset").alias("avg_cost_offset")
)

The next step would try to process and aggregate the power load data, extracting key load parameters (specifically real "pd" and reactive "qd" power demands) per snapshot.

In [14]:
# Extracting load data
load_df = df.select(
    "filename",
    col("grid.nodes.load").alias("load_arrays")  
).select(
    "filename",
    explode("load_arrays").alias("load")  
).select(
    "filename",
    col("load")[0].alias("pd"),  
    col("load")[1].alias("qd")
)

load_df.show(5)


+--------------------+-------------------+--------------------+
|            filename|                 pd|                  qd|
+--------------------+-------------------+--------------------+
|gs://bucketseattl...|0.32349711373191603|  0.0541399530109636|
|gs://bucketseattl...| 0.8572351780984674|   0.253666451943297|
|gs://bucketseattl...| 0.5207840155565728| 0.12156684298552808|
|gs://bucketseattl...| 0.7133438352218501|0.039086170526506626|
|gs://bucketseattl...| 0.6837468540898303| 0.13888321084001698|
+--------------------+-------------------+--------------------+
only showing top 5 rows



In [15]:
# Aggregate load features per snapshot
load_agg = load_df.groupBy("filename").agg(
    count("*").alias("num_loads"),
    spark_sum("pd").alias("total_pd")
)

Similarly, the next step would try to extract and aggregate transmission line features.

In [16]:
# Extract transmission line features from grid.edges.ac_line.features
ac_line_df = df.select(
    "filename",
    explode("grid.edges.ac_line.features").alias("ac_line")
).select(
    "filename",
    col("ac_line")[2].alias("br_r"),
    col("ac_line")[3].alias("br_x"),
    col("ac_line")[4].alias("rate_a"),
    col("ac_line")[5].alias("rate_b"),
    col("ac_line")[6].alias("rate_c")
)

ac_line_df.show(5)

+--------------------+----------+----------+----------+---------+------------------+
|            filename|      br_r|      br_x|    rate_a|   rate_b|            rate_c|
+--------------------+----------+----------+----------+---------+------------------+
|gs://bucketseattl...|0.01340085|0.01340085| 0.0154525|0.0792528|            2.3994|
|gs://bucketseattl...| 0.0065814| 0.0065814|0.00358046|0.0258456|            2.6003|
|gs://bucketseattl...| 0.0065814| 0.0065814|0.00358046|0.0258456|            2.6003|
|gs://bucketseattl...| 0.0062579| 0.0062579|0.00319977|0.0219502|2.5658999999999996|
|gs://bucketseattl...|0.00557485|0.00557485|0.00696197|0.0360534|            2.1877|
+--------------------+----------+----------+----------+---------+------------------+
only showing top 5 rows



In [17]:
# Aggregate transformer features per snapshot
ac_line_agg = ac_line_df.groupBy("filename").agg(
    avg("br_r").alias("br_r_mean"),
    avg("br_x").alias("br_x_mean"),
    spark_sum("rate_a").alias("rate_a_sum"),
    F.min("rate_b").alias("rate_b_min"),
    F.max("rate_c").alias("rate_c_max")
)

Similarly, extract and aggregate the Transformer features

In [18]:
# Extract transformer features from grid.edges.transformer.features
trans_df = df.select(
    "filename",
    explode("grid.edges.transformer.features").alias("trans")
).select(
    "filename",
    col("trans")[2].alias("br_r"),
    col("trans")[3].alias("br_x"),
    col("trans")[4].alias("rate_a"),
    col("trans")[7].alias("tap")
)

trans_df.show(5)

+--------------------+----------+---------+------+-------+
|            filename|      br_r|     br_x|rate_a|    tap|
+--------------------+----------+---------+------+-------+
|gs://bucketseattl...|5.36254E-4|0.0293108| 3.905| 1.0125|
|gs://bucketseattl...|5.05881E-4|0.0243569|5.0692|1.03125|
|gs://bucketseattl...|5.41271E-4|0.0336934|4.6108|    1.1|
|gs://bucketseattl...|5.41271E-4|0.0336934|4.6108|    1.1|
|gs://bucketseattl...|5.41271E-4|0.0336934|4.6108|    1.1|
+--------------------+----------+---------+------+-------+
only showing top 5 rows



In [19]:
# Aggregate transformer features per snapshot
trans_agg = trans_df.groupBy("filename").agg(
    avg("br_r").alias("trans_br_r_mean"),
    avg("br_x").alias("trans_br_x_mean"),
    spark_sum("rate_a").alias("trans_rate_a_sum"),
    avg("tap").alias("tap_mean")
)

Similarly, extract the objetive value (total cost).

In [20]:
# Extract the objective value directly from metadata
obj_df = df.select(
    "filename",
    col("metadata.`objective`").alias("total_cost") 
)


obj_df.show(5)

+--------------------+------------------+
|            filename|        total_cost|
+--------------------+------------------+
|gs://bucketseattl...| 443934.8106702195|
|gs://bucketseattl...|465533.45792886155|
|gs://bucketseattl...| 462092.4328079719|
|gs://bucketseattl...| 453110.9782805654|
|gs://bucketseattl...|452267.02218503354|
+--------------------+------------------+
only showing top 5 rows



Then, we will merge aggregated data from different sources and ensures data consistency by checking for duplicate entries. Here are the main steps:
  
- Joining aggregated data from generator, load, transmission line, transformer and objective value based on "filename" into a single DataFrame.
- Validating the aggregated results
- Checking for duplicate filenames

In [21]:
# Use inner joins on 'filename' (the snapshot key) to combine all real values
agg_df = gen_agg.join(load_agg, "filename", "inner") \
                .join(ac_line_agg, "filename", "inner") \
                .join(trans_agg, "filename", "inner") \
                .join(obj_df, "filename", "inner")

# Validate the result
print("Total number of rows after aggregation:", agg_df.count())
agg_df.show(5, truncate=False)  # Display the first 5 rows without truncation

# Check for duplicate filenames
from pyspark.sql import functions as F

duplicate_check = agg_df.groupBy("filename").agg(
    F.count("*").alias("count")
).filter(F.col("count") > 1)

if duplicate_check.count() == 0:
    print("All filenames are unique")
else:
    print("Duplicate filenames exist! Data consistency needs to be checked")


Total number of rows after aggregation: 100
+--------------------------------------------------------------------------------------------------------------+--------------+------------------+------+------------------+-----------------+------------------+---------+------------------+--------------------+--------------------+------------------+----------+----------+---------------------+-------------------+-----------------+------------------+------------------+
|filename                                                                                                      |num_generators|total_pg          |avg_vg|avg_cost_squared  |avg_cost_linear  |avg_cost_offset   |num_loads|total_pd          |br_r_mean           |br_x_mean           |rate_a_sum        |rate_b_min|rate_c_max|trans_br_r_mean      |trans_br_x_mean    |trans_rate_a_sum |tap_mean          |total_cost        |
+--------------------------------------------------------------------------------------------------------------+----

### Address missing data
Author: Ewan Long

Date: 03/24/25

In this task, I will do the data quality checks, and manages any potential missing data, and enforces consistency for a prepared aggregated dataset.

___

Firstly, Calculates the count and proportion of missing values for each column

In [22]:
# Count missing values per column
missing_count_df = agg_df.select([
    F.count(F.when(F.isnull(c), c)).alias(c) for c in agg_df.columns
])

# Total rows in DataFrame
total_rows = agg_df.count()
total_row_df = spark.createDataFrame([(total_rows,)], ["total_rows"])

# Combine counts and total row number via cross join
missing_stats = missing_count_df.crossJoin(total_row_df)

# Compute missing ratios
missing_stats = missing_stats.select(
    # Original missing counts
    *[F.col(c) for c in agg_df.columns],
    # Missing ratio per column
    *[(F.col(c) / F.col("total_rows")).alias(f"{c}_missing_ratio") for c in agg_df.columns]
)


Show the overview of missing values and their ratios for each column in the aggregated DataFrame

In [23]:
# Display missing value statistics clearly
print(f"Total rows: {total_rows}")
missing_stats.show(vertical=True, truncate=False)  
# using vertical=True improves readability for large number of columns

Total rows: 100
-RECORD 0-----------------------------
 filename                       | 0   
 num_generators                 | 0   
 total_pg                       | 0   
 avg_vg                         | 0   
 avg_cost_squared               | 0   
 avg_cost_linear                | 0   
 avg_cost_offset                | 0   
 num_loads                      | 0   
 total_pd                       | 0   
 br_r_mean                      | 0   
 br_x_mean                      | 0   
 rate_a_sum                     | 0   
 rate_b_min                     | 0   
 rate_c_max                     | 0   
 trans_br_r_mean                | 0   
 trans_br_x_mean                | 0   
 trans_rate_a_sum               | 0   
 tap_mean                       | 0   
 total_cost                     | 0   
 filename_missing_ratio         | 0.0 
 num_generators_missing_ratio   | 0.0 
 total_pg_missing_ratio         | 0.0 
 avg_vg_missing_ratio           | 0.0 
 avg_cost_squared_missing_ratio | 0.0 
 avg_cost

### Creating Pipelines and Doing Experiments
  
Author: Donna Nguyen
  
Date: 03/24/25
  
Create baseline pipelines and do experiments on your data:

Set a blind test set that is never seen during training (10% of data)
Report evaluation metrics over the training set
Report evaluation metrics over blind test dataset
Create a baseline model using linear regression
Discussion of experimental results on training and testing dataset

In [ ]:
from pyspark.ml.feature import VectorAssembler, MinMaxScaler
from pyspark.ml.regression import LinearRegression
from pyspark.ml import Pipeline
from pyspark.ml.evaluation import RegressionEvaluator

# Define numerical columns (exclude total_cost and filename)
numerical_columns = [col for col in agg_df.columns if col not in ["total_cost", "filename"]]

# Assemble features into a single vector column
assembler = VectorAssembler(inputCols=numerical_columns, outputCol="unscaled_features")

# Apply MinMaxScaler for feature scaling
scaler = MinMaxScaler(inputCol="unscaled_features", outputCol="features")

# Define Linear Regression model
lr = LinearRegression(featuresCol="features", labelCol="total_cost", maxIter=100, regParam=0.1)

# Create ML pipeline
pipeline = Pipeline(stages=[assembler, scaler, lr])

# Split data into training (90%) and blind test set (10%)
train_data, test_data = agg_df.randomSplit([0.9, 0.1], seed=42)

# Train the model
model = pipeline.fit(train_data)

# Make predictions on train and test sets
train_predictions = model.transform(train_data)
test_predictions = model.transform(test_data)

# Define evaluator for RMSE and R²
rmse_evaluator = RegressionEvaluator(labelCol="total_cost", predictionCol="prediction", metricName="rmse")
r2_evaluator = RegressionEvaluator(labelCol="total_cost", predictionCol="prediction", metricName="r2")

# Compute RMSE and R² for training set
train_rmse = rmse_evaluator.evaluate(train_predictions)
train_r2 = r2_evaluator.evaluate(train_predictions)

# Compute RMSE and R² for test set
test_rmse = rmse_evaluator.evaluate(test_predictions)
test_r2 = r2_evaluator.evaluate(test_predictions)

# Print model performance
print(f"Training Set -> RMSE: {train_rmse:.4f}, R²: {train_r2:.4f}")
print(f"Test Set -> RMSE: {test_rmse:.4f}, R²: {test_r2:.4f}")

# Show sample predictions
test_predictions.select("total_cost", "prediction").show(5, truncate=False)


Training Set -> RMSE: 270.7873, R²: 0.9983
Test Set -> RMSE: 521.4989, R²: 0.9969
+------------------+-----------------+
|total_cost        |prediction       |
+------------------+-----------------+
|454043.8197398794 |454131.215759768 |
|452002.79294327396|452704.7248362681|
|444792.7982146128 |444610.5818190558|
|455445.203582367  |455317.9064252551|
|436232.22792202106|435323.1401192636|
+------------------+-----------------+
only showing top 5 rows



In [25]:
# Define evaluators for RMSE, R², and MSE
rmse_evaluator = RegressionEvaluator(labelCol="total_cost", predictionCol="prediction", metricName="rmse")
r2_evaluator = RegressionEvaluator(labelCol="total_cost", predictionCol="prediction", metricName="r2")
mse_evaluator = RegressionEvaluator(labelCol="total_cost", predictionCol="prediction", metricName="mse")

# Compute metrics for training set
train_rmse = rmse_evaluator.evaluate(train_predictions)
train_r2 = r2_evaluator.evaluate(train_predictions)
train_mse = mse_evaluator.evaluate(train_predictions)

# Compute metrics for test set
test_rmse = rmse_evaluator.evaluate(test_predictions)
test_r2 = r2_evaluator.evaluate(test_predictions)
test_mse = mse_evaluator.evaluate(test_predictions)

# Print model performance
print(f"Training Set -> RMSE: {train_rmse:.4f}, MSE: {train_mse:.4f}, R²: {train_r2:.4f}")
print(f"Test Set -> RMSE: {test_rmse:.4f}, MSE: {test_mse:.4f}, R²: {test_r2:.4f}")


Training Set -> RMSE: 270.7873, MSE: 73325.7589, R²: 0.9983
Test Set -> RMSE: 521.4989, MSE: 271961.0970, R²: 0.9969


### Discussions

The baseline linear regression model demonstrated strong predictive performance on both the training and test datasets. A high R² (close to 1) indicate that the variance in total_cost is well-explained by the provided quantitative features. The test RMSE is higher than the training RMSE, indicating a slight increase in prediction error on unseen data. However, the high R² value (0.9969) suggests that the model still generalizes well. The prediction values align quite closely with the actual cost so far. But this is only an experiment on a small sample though. The slight performance drop from training to test set suggests some degree of overfitting, though the effect seems minimal. Further improvement could involve regularization tuning (adjusting regParam), more Feature engineering or just trying other models for comparison.

